hw/drivers/debounce/src/debounce.c

```diff
@@ -0,0 +1,128 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "debounce/debounce.h"
+#include "syscfg/syscfg.h"
+#include <string.h>
```

Comment:
As a rule, it is better to place syscfg.h before including any header files from within mynewt; syscfg.h contents influence conditional parts in header files.

Reply:
Shouldn't those header files then include syscfg themselves ?
Includes should be alphabetically ordered, makes it a lot easier to figure out what is included and what isn't.

Reply:
Most uses of syscfg involve the `MYNEWT_VAL()` macro.  If a header tries to use `MYNEWT_VAL()` before `syscfg.h` has been included, you'll get a build error.  So, in these cases, it is not terribly important that the `syscfg.h` include go at the top.

However, in rare cases, code needs to check if a syscfg macro has been defined at all (e.g., `log/full/include/log/log.h` uses `#ifndef MYNEWT_VAL_LOG_LEVEL`).  This is needed when you must distinguish an undefined setting from a setting that is defined as `0`.  In these cases, forgetting to included `syscfg.h` won't trigger a build error and can lead to lots of confusion.  My opinion is that preventing this issue is more important than the ability to quickly see which headers are being included, so I agree with @mkiiskila that `syscfg.h` should always be included before any other Mynewt headers.

Reply:
my point was, that header files need to be self contained - they need to include whatever they need and not rely on the fact that another file that included that header includes another header file before that.

So unless this header file needs the contents of syscfg.h in order to determine if it should include another header file or not - it doesn't matter where in the include list it is.

As to promoting including a specific header always first or last (or 3rd), I don't consider that a very good coding style. It encourages and promotes badly formed header and source files. It would be more prudent to fix all headers and sources which rely on syscfg.h but don't include it.

Reply:
I agree that headers should be self-contained.  Including `syscfg.h` early is just protection in case you forget.

Re: good coding style - join the debate :).  I think you'll find everyone has his own opinion on this particular issue (e.g., https://stackoverflow.com/questions/2762568/c-c-include-file-order-best-practices).  This is certainly not a huge issue, but my feeling is that including `syscfg.h` first is a practical solution to an actual problem.  The ability to quickly see which headers are included seems less useful.